modules/gapi/src/backends/fluid/gfluidcore.cpp

```diff
@@ -58,6 +58,339 @@ static inline DST add(SRC1 x, SRC2 y)
+                                                reinterpret_cast<ushort*>(out), length, x);
+
+    }
+    else if (std::is_same<DST, short>::value)
+    {
+        absdiff_simd_sametype<short, v_int16>(reinterpret_cast<const short*>(in1),
+                                              reinterpret_cast<const short*>(in2),
+                                              reinterpret_cast<short*>(out), length, x);
+    }
+    else if (std::is_same<DST, float>::value)
+    {
+        absdiff_simd_sametype<float, v_float32>(reinterpret_cast<const float*>(in1),
+                                                reinterpret_cast<const float*>(in2),
+                                                reinterpret_cast<float*>(out), length, x);
+    }
+}
+
+template<typename T, typename VT>
+static inline void add_simd_sametype(const T in1[], const T in2[], T out[], int length, int& x)
+{
+    constexpr int nlanes = VT::nlanes;
+    for (; x < length; )
+    {
+        for (; x <= length - nlanes; x += nlanes)
+        {
+            VT a = vx_load(&in1[x]);
+            VT b = vx_load(&in2[x]);
+            vx_store(&out[x], a + b);
+        }
+
+        if (x < length && length >= nlanes)
+        {
+            x = length - nlanes;
+        }
+    }
+}
+
+template<typename SRC, typename DST>
+static inline void add_simd(const SRC in1[], const SRC in2[], DST out[], int length, int& x)
+{
+    if (std::is_same<DST, SRC>::value)
+    {
+        if (std::is_same<DST, uchar>::value)
+        {
+            add_simd_sametype<uchar, v_uint8>(reinterpret_cast<const uchar*>(in1),
+                                              reinterpret_cast<const uchar*>(in2),
+                                              reinterpret_cast<uchar*>(out), length, x);
+        }
+        else if (std::is_same<DST, short>::value)
+        {
+            add_simd_sametype<short, v_int16>(reinterpret_cast<const short*>(in1),
+                                              reinterpret_cast<const short*>(in2),
+                                              reinterpret_cast<short*>(out), length, x);
+        }
+        else if (std::is_same<DST, float>::value)
+        {
+            add_simd_sametype<float, v_float32>(reinterpret_cast<const float*>(in1),
+                                                reinterpret_cast<const float*>(in2),
+                                                reinterpret_cast<float*>(out), length, x);
+        }
+    }
+    else if (std::is_same<SRC, uchar>::value && std::is_same<DST, float>::value)
+    {
+        constexpr int nlanes = v_float32::nlanes;
+        for (; x < length; )
+        {
+            for (; x <= length - nlanes; x += nlanes)
+            {
+                v_uint32 a = vx_load_expand_q(reinterpret_cast<const uchar*>(&in1[x]));
+                v_uint32 b = vx_load_expand_q(reinterpret_cast<const uchar*>(&in2[x]));
+                vx_store(reinterpret_cast<float*>(&out[x]), v_cvt_f32(a + b));
```

Comment:
Would it be faster in theory if all these `reinterpret_cast`s were made outside of the cycle?

Reply:
No, it's not good solution. Also it don't give speedup.

Reply:
ok